airflow/providers/openlineage/extractors/base.py

```diff
@@ -0,0 +1,133 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from __future__ import annotations
+
+import json
+from abc import ABC, abstractmethod
+from urllib.parse import parse_qsl, urlencode, urlparse, urlunparse
+
+from attrs import Factory, define
+
+from airflow.utils.log.logging_mixin import LoggingMixin
+from openlineage.client.facet import BaseFacet
+from openlineage.client.run import Dataset
+
+
+@define
+class OperatorLineage:
+    """Structure returned from lineage extraction."""
+
+    inputs: list[Dataset] = Factory(list)
+    outputs: list[Dataset] = Factory(list)
+    run_facets: dict[str, BaseFacet] = Factory(dict)
+    job_facets: dict[str, BaseFacet] = Factory(dict)
+
+
+class BaseExtractor(ABC, LoggingMixin):
+    """
+    Abstract base extractor class.
+
+    This is used mostly to maintain support for custom extractors.
+    """
+
+    _allowed_query_params: list[str] = []
+
+    def __init__(self, operator):
+        super().__init__()
+        self.operator = operator
+        self.patch()
+
+    def patch(self):
+        # Extractor should register extension methods or patches to operator here
+        pass
+
+    @classmethod
+    def get_operator_classnames(cls) -> list[str]:
+        """
+        Implement this method returning list of operators that extractor works for.
+        Particularly, in Airflow 2 some operators are deprecated and simply subclass the new
+        implementation, for example BigQueryOperator:
+        https://github.com/apache/airflow/blob/main/airflow/contrib/operators/bigquery_operator.py
+        The BigQueryExtractor needs to work with both of them.
+        :return:
+        """
+        raise NotImplementedError()
+
+    def validate(self):
+        assert self.operator.__class__.__name__ in self.get_operator_classnames()
```

Comment:
This would mess up with mapped operator I think

```suggestion
        assert self.operator.task_type in self.get_operator_classnames()
```

Reply:
Done